### PR TITLE
Added debug to equalWithinTolerance

### LINKED ISSUE #1371: Add debug to equalWithinTolerance
### Description of the task
Add debug print outs to equalWithinTolerance so we can see why a test failed

### Acceptance criteria

- [ ] Add debug to equalWithinTolerance

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Added debug print outs to equalWithinTolerance

### Testing Done
CI

### Resolved Issues
Resolves #1371

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by akhilveeraghanta — COMMENTED** [review]
lgtm, although you may be able to get away without using a `debug` flag because these are test utils and it doesn't hurt to print every time

**Comment by jonathanlew** [comment]
> lgtm, although you may be able to get away without using a `debug` flag because these are test utils and it doesn't hurt to print every time

fair point, but it does get overly verbose since equalWithinTolerance is implemented somewhat recursively, i.e. implementations of polygon calls point which calls double

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Personally not a fan of adding a debug flag. I don't think it's super obvious to people they will need to set it.

Can't believe I didn't think of this before. Why don't we implement custom gtest "predicate checking" functions? That way they can act a bit more like existing gtest asserts, and have mechanisms to only print info on failure.

https://stackoverflow.com/a/45247291

https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#using-a-function-that-returns-an-assertionresult

I think it should just require minor changes to the existing `equalWithinTolerance` functions, but should give us all the benefits we're looking for.

File: src/software/test_util/test_util.cpp [hunk 1]
```diff
@@ -104,35 +106,84 @@ namespace Test
...
+                                        double tolerance, bool debug)
     {
         auto ppts1 = poly1.getPoints();
         auto ppts2 = poly2.getPoints();
-        return std::equal(ppts1.begin(), ppts1.end(), ppts2.begin(),
-                          [tolerance](const Point &p1, const Point &p2) {
-                              return equalWithinTolerance(p1, p2, tolerance);
-                          });
+        if (std::equal(ppts1.begin(), ppts1.end(), ppts2.begin(),
+                       [tolerance](const Point &p1, const Point &p2) {
```
Comment (EvanMorcom) [inline src/software/test_util/test_util.cpp:115]: Does this only work of the Points that make up the polygon are in the same order?

**Review by EvanMorcom — COMMENTED** [review]
LGTM 🐕 🐈 

Just one comment

File: src/software/test_util/test_util.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/test_util/test_util.cpp:115]: yes

**Comment by jonathanlew** [comment]
> Personally not a fan of adding a debug flag. I don't think it's super obvious to people they will need to set it.
> 
> Can't believe I didn't think of this before. Why don't we implement custom gtest "predicate checking" functions? That way they can act a bit more like existing gtest asserts, and have mechanisms to only print info on failure.
> 
> https://stackoverflow.com/a/45247291
> 
> https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#using-a-function-that-returns-an-assertionresult
> 
> I think it should just require minor changes to the existing `equalWithinTolerance` functions, but should give us all the benefits we're looking for.

@MathewMacDougall I've made the switch

**Review by akhilveeraghanta — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]
very slick, thanks for changing :+1:

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**